demos/https/iot_demo_https_s3_upload_async.c

```diff
@@ -0,0 +1,538 @@
+     * acquire everything following in IOT_DEMO_HTTPS_PRESIGNED_PUT_URL. */
+    reqConfig.pPath = pPath;
+    reqConfig.pathLen = strlen( pPath );
+    reqConfig.pHost = pAddress;
+    reqConfig.hostLen = addressLen;
+
+    /* The PUT method is used to upload an object to S3 because it is simpler than using POST. POST requires AWS S3
+     * access specific header fields in the body of the message in a multipart/form-data encoded message. See
+     * https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html for more information about POST object. See
+     * https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html for more information about PUT object. */
+    reqConfig.method = IOT_HTTPS_METHOD_PUT;
+    reqConfig.isNonPersistent = false;
+    reqConfig.userBuffer.pBuffer = _pReqUserBuffer;
+    reqConfig.userBuffer.bufferLen = sizeof( _pReqUserBuffer );
+    reqConfig.isAsync = true;
+    reqConfig.u.pAsyncInfo = &asyncInfo;
+
+    respConfig.userBuffer.pBuffer = _pRespUserBuffer;
+    respConfig.userBuffer.bufferLen = sizeof( _pRespUserBuffer );
+    respConfig.pSyncInfo = NULL;
+
+    /* Create the semaphore for waiting for the whole upload response to finish. */
+    uploadFinishedSemCreated = IotSemaphore_Create( &( _uploadFinishedSem ), 0 /* Initial value. */, 1 /* Max value. */ );
+
+    if( uploadFinishedSemCreated == false )
+    {
+        IotLogError( "Failed to create a semaphore to wait for the response to finish." );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Initialize the HTTPS library. */
+    httpsClientStatus = IotHttpsClient_Init();
+
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "An error occurred initializing the HTTPS library. Error code: %d", httpsClientStatus );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Initialize the request to initialize internal request context, write the HTTP request first line to the
+     * _pReqUserBuffer header space, and write headers User-Agent and Host. */
+    httpsClientStatus = IotHttpsClient_InitializeRequest( &reqHandle, &reqConfig );
+
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "An error occurred in IotHttpsClient_InitializeRequest() with error code: %d", httpsClientStatus );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Connect to S3. */
+    for( connIndex = 1; connIndex <= IOT_DEMO_HTTPS_CONNECTION_NUM_RETRY; connIndex++ )
+    {
+        httpsClientStatus = IotHttpsClient_Connect( &connHandle, &connConfig );
+
+        if( ( httpsClientStatus == IOT_HTTPS_CONNECTION_ERROR ) &&
+            ( connIndex < IOT_DEMO_HTTPS_CONNECTION_NUM_RETRY ) )
+        {
+            IotLogError( "Failed to connect to the S3 server, retrying after %d ms.",
+                         IOT_DEMO_HTTPS_CONNECTION_RETRY_WAIT_MS );
+            IotClock_SleepMs( IOT_DEMO_HTTPS_CONNECTION_RETRY_WAIT_MS );
```

Comment:
Does this call pend the task in the same way that the async send does below? Just a sanity check.

Reply:
This call will have the main demo task wait a few ms before trying to connect again.
IotHttpsClient_SendAsync() will not block the demo task.